cert.py

```diff
@@ -144,17 +155,32 @@ def _get_http_challenge(ctx, authzr):
     ctx.fail('no acceptable challenge type found; only HTTP01 supported')
 
 
-def _store_webroot_validation(webroot, challb, val):
+def _store_webroot_validation(ctx, remote, webroot, challb, val):
     logger.info('storing validation of %s' % webroot)
-    try:
-        os.makedirs(os.path.join(webroot, challb.URI_ROOT_PATH), 0o755)
-    except OSError as e:
-        if e.errno != errno.EEXIST:
-            raise
-
-    with open(os.path.join(webroot, challb.path.strip('/')), 'wb') as outf:
-        logger.info('storing validation to %s' % outf.name)
-        outf.write(b(val))
+    chall_path = os.path.join(webroot, challb.path.strip('/'))
+    if not remote:
+        try:
+            os.makedirs(os.path.join(webroot, challb.URI_ROOT_PATH), 0o755)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+        with open(chall_path, 'wb') as outf:
+            logger.info('storing validation to %s' % outf.name)
+            outf.write(b(val))
+    else:
+        logger.info('storing validation to %s:%s' % (remote[1], chall_path))
+        ssh = pxssh()
+        ssh.force_password = True
+        ssh.SSH_OPTS = "-o 'NumberOfPasswordPrompts=1'"
+        user, host, port, password = remote
+        try:
+            if ssh.login(host, user, password, port=port):
+                ssh.sendline('echo "%s" > %s' % (b(val), chall_path))
```

Comment:
I do not see the difference between running remote `echo` command via SSH and writing directly to file as the script is doing it now.

If it needs to be even more restrictive I could check for unwanted characters in filename so that potential attacker can not navigate with `../`, etc. Also, account can use `rbash` that allows only `echo` command. Considering that `.well-known` folder is owned by `wile` user and `wile` user does not have home folder created he could only `echo` to challenge file and if breached possibly write to `tmp`.

But all of the mentioned problems are also existent in current way of doing things? Script still needs to be executed as some user, `wile` for example.

Also, I'm considering switching from `pexpect` (`pxssh`) to `paramiko`.